test/core/client_channel/lb_policy/pick_first_test.cc

```diff
@@ -39,6 +49,70 @@ class PickFirstTest : public LoadBalancingPolicyTest {
  protected:
   PickFirstTest() : lb_policy_(MakeLbPolicy("pick_first")) {}
 
+  static RefCountedPtr<LoadBalancingPolicy::Config> MakePickFirstConfig(
+      bool shuffle_address_list) {
+    return MakeConfig(Json::FromArray({Json::FromObject({{
+        "pick_first",
+        Json::FromObject(
+            {{"shuffleAddressList", Json::FromBool(shuffle_address_list)}}),
+    }})}));
+  }
+
+  // Gets order the addresses are being picked. Return type is void so
+  // assertions can be used
+  void GetOrderAddressesArePicked(
+      absl::Span<const absl::string_view> addresses,
+      std::vector<absl::string_view>* out_address_order) {
+    // Construct a map of subchannel to address.
+    // We will remove entries as each subchannel starts to connect.
+    std::map<SubchannelState*, absl::string_view> subchannels;
+    for (auto address : addresses) {
+      auto* subchannel = FindSubchannel(
+          address, ChannelArgs().Set(GRPC_ARG_INHIBIT_HEALTH_CHECKING, true));
+      ASSERT_NE(subchannel, nullptr);
+      subchannels.emplace(subchannel, address);
+    }
+    // Now process each subchannel in the order in which pick_first tries it.
+    while (!subchannels.empty()) {
+      // Find the subchannel that is being attempted.
+      SubchannelState* subchannel = nullptr;
+      for (const auto& p : subchannels) {
+        if (p.first->ConnectionRequested()) {
+          out_address_order->push_back(p.second);
+          subchannel = p.first;
+          break;
+        }
+      }
+      ASSERT_NE(subchannel, nullptr);
+      // The subchannel reports CONNECTING.
+      subchannel->SetConnectivityState(GRPC_CHANNEL_CONNECTING);
+      // If this is the first subchannel being attempted, expect a CONNECTING
+      // update.
+      if (subchannels.size() == addresses.size()) {
+        ExpectConnectingUpdate();
+      }
+      if (subchannels.size() > 1) {
+        // Not the last subchannel in the list.  Connection attempt should fail.
+        subchannel->SetConnectivityState(
+            GRPC_CHANNEL_TRANSIENT_FAILURE,
+            absl::UnavailableError("failed to connect"));
+        subchannel->SetConnectivityState(GRPC_CHANNEL_IDLE);
+      } else {
```

Comment:
I think we should remove this `else` block.  We can just leave the last subchannel trying to connect, since the policy will stop using it as soon as we deliver a new address list.

Reply:
I left it so we test that the addresses are not reshuffled.

Reply:
I don't think this is going to work reliably.  The test is going to be flaky.

The problem is that when we call `ExpectStateAndQueuingPicker(GRPC_CHANNEL_IDLE)` for the last subchannel, doing the pick will trigger the LB policy to exit idle state and start trying to connect again.  It will start with the first subchannel in the *current* order of the addresses, so `ConnectionRequested()` will return true for that subchannel even before we send a new address list update to the LB policy, which will cause it to shuffle again.  So the next time we call `GetOrderAddressesArePicked()`, we may see the connection attempts in a different order than the new shuffled order.

For example, consider the following sequence of events:
1. We send the LB policy an address update containing [A, B, C].  The LB policy shuffles to the order [C, B, A].
1. On the first call to `GetOrderAddressesArePicked()`, we see the shuffled order [C, B, A].  We fail the connection attempts to C and B, and then we succeed connecting to A.  Then we have A report IDLE, which causes the policy to report IDLE.  Now we request a pick, which triggers the policy to start trying to connect again.  It will request a connection on C, since that's the first address in the current order.  At this point, the subchannel for C will return true for `ConnectionRequested()`, but we haven't yet checked it.
2. Now we send the LB policy a new address list update.  This time, it shuffles the addresses into order [B, A, C].  It will now immediately drop the old subchannel list and switch to the new one, and it will start connecting to address B.  This means the subchannel for B will return true for `ConnectionRequested()`.
6. In the test, we now call `GetOrderAddressesArePicked()` again.  At this point, the subchannels for both addresses B and C will return true for `ConnectionRequested()`.  We want the test to see B first, since that's the first entry in the shuffled address list, but it may happen to see C first instead, thus returning a different order than what the LB policy is actually using.  (The order in which we check the addresses here depends on the addresses that happen to be used to allocate the `SubchannelState` objects, since that address is the key of the map on line 69.)

To solve this, I think you need to add a parameter to this function indicating whether we want to finish the connection attempt pass.  In the case where we are going to send another address update after this one, we should avoid doing a pick after the LB policy reports IDLE, so that it does not start trying to connect again.  In other words, we want something like this on line 110:

```
if (exit_idle_after_connection_fails) {
  ExpectStateAndQueuingPicker(GRPC_CHANNEL_IDLE);
} else {
  ExpectState(GRPC_CHANNEL_IDLE);
}
```

But this also means that we need to trigger the LB policy to exit idle *after* we deliver the subsequent address list update to it.  So we'll probably need to add code at the top of this function to hop into the `WorkSerializer` and call the LB policy's `ExitIdleLocked()` method.

Reply:
Done